components/swim/src/member.rs

```diff
@@ -0,0 +1,706 @@
+    pub fn new() -> Member {
+        let mut proto_member = ProtoMember::new();
+        proto_member.set_id(Uuid::new_v4().simple().to_string());
+        proto_member.set_incarnation(0);
+        Member { proto: proto_member }
+    }
+
+    /// Returns the socket address of this member.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if the address is un-parseable. In practice, it shouldn't be
+    /// un-parseable, since its set from the inbound socket directly.
+    pub fn socket_address(&self) -> SocketAddr {
+        match self.get_address().parse() {
+            Ok(addr) => addr,
+            Err(e) => {
+                panic!("Cannot parse member {:?} address: {}", self, e);
+            }
+        }
+    }
+}
+
+impl Deref for Member {
+    type Target = ProtoMember;
+
+    fn deref(&self) -> &ProtoMember {
+        &self.proto
+    }
+}
+
+impl DerefMut for Member {
+    fn deref_mut(&mut self) -> &mut ProtoMember {
+        &mut self.proto
+    }
+}
+
+impl From<ProtoMember> for Member {
+    fn from(member: ProtoMember) -> Member {
+        Member { proto: member }
+    }
+}
+
+impl<'a> From<&'a ProtoMember> for Member {
+    fn from(member: &'a ProtoMember) -> Member {
+        Member { proto: member.clone() }
+    }
+}
+
+impl From<SocketAddr> for Member {
+    fn from(socket: SocketAddr) -> Member {
+        let mut member = Member::new();
+        member.set_address(format!("{}", socket));
+        member
+    }
+}
+
+impl From<Member> for RumorKey {
+    fn from(member: Member) -> RumorKey {
+        RumorKey::new("member", member.get_id())
+    }
+}
+
+impl<'a> From<&'a Member> for RumorKey {
+    fn from(member: &'a Member) -> RumorKey {
+        RumorKey::new("member", member.get_id())
+    }
+}
+
+impl<'a> From<&'a &'a Member> for RumorKey {
+    fn from(member: &'a &'a Member) -> RumorKey {
+        RumorKey::new("member", member.get_id())
+    }
+}
+
+// This is a Uuid type turned to a string
+pub type UuidSimple = String;
+
+/// Tracks lists of members, their health, and how long they have been suspect.
+#[derive(Debug, Clone)]
+pub struct MemberList {
+    members: Arc<RwLock<HashMap<UuidSimple, Member>>>,
+    health: Arc<RwLock<HashMap<UuidSimple, Health>>>,
+    suspect: Arc<RwLock<HashMap<UuidSimple, SteadyTime>>>,
+}
+
+impl MemberList {
+    /// Creates a new, empty, MemberList.
+    pub fn new() -> MemberList {
+        MemberList {
+            members: Arc::new(RwLock::new(HashMap::new())),
+            health: Arc::new(RwLock::new(HashMap::new())),
+            suspect: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Inserts a member into the member list with the given health.
+    pub fn insert(&self, member: Member, health: Health) -> bool {
+        let share_rumor: bool;
+        let mut start_suspicion: bool = false;
+        let mut stop_suspicion: bool = false;
+
+        // If we have an existing member record..
+        if let Some(current_member) = self.members
```

Comment:
Yeah, it's gotten pretty hairy. It's fairly similar to the logic that's in the existing implementation, which also look similar. It's made worse by needing to not contend for the locks.